File: laundry-reservation/python/src/laundry_reservation/reservation_service.py
from __future__ import annotations

import uuid
from datetime import datetime

from .customer import Customer
from .email_notifier import EmailNotifier
from .exceptions import DuplicateReservationError
from .machine_api import MachineApi
from .machine_selector import MachineSelector
from .pin_generator import PinGenerator
from .reservation import Reservation
from .reservation_repository import ReservationRepository
from .sms_notifier import SmsNotifier

MAX_FAILED_ATTEMPTS = 5


class ReservationService:
    def __init__(
        self,
        repository: ReservationRepository,
        email_notifier: EmailNotifier,
        sms_notifier: SmsNotifier,
        machine_api: MachineApi,
        pin_generator: PinGenerator,
        machine_selector: MachineSelector,
    ) -> None:
        self._repository = repository
        self._email_notifier = email_notifier
        self._sms_notifier = sms_notifier
        self._machine_api = machine_api
        self._pin_generator = pin_generator
        self._machine_selector = machine_selector
        self._failure_counts: dict[int, int] = {}

    def create_reservation(self, slot: datetime, customer: Customer) -> Reservation:
        existing = self._repository.find_active_by_customer_email(customer.email)
        if existing is not None:
            raise DuplicateReservationError(
                f"Customer '{customer.email}' already has an active reservation."
            )

        reservation_id = str(uuid.uuid4())
        machine_number = self._machine_selector.select_available()
        pin = self._pin_generator.generate()
        reservation = Reservation(reservation_id, slot, machine_number, customer, pin)

        self._repository.save(reservation)

        self._email_notifier.send(
            customer.email,
            "Wunda Wash Reservation Confirmation",
            f"Reservation {reservation_id}: Machine {machine_number}, PIN {pin:05d}",
        )

        self._machine_api.lock(reservation_id, machine_number, slot, pin)

        return reservation

    def claim_reservation(self, machine_number: int, pin: int) -> bool:
        reservation = self._repository.find_active_by_machine_number(machine_number)
        if reservation is None:
            return False

        if reservation.pin == pin:
            reservation.mark_used()
            self._repository.save(reservation)
            self._machine_api.unlock(machine_number, reservation.id)
            self._failure_counts.pop(machine_number, None)
            return True

        failures = self._failure_counts.get(machine_number, 0) + 1
        self._failure_counts[machine_number] = failures

        if failures >= MAX_FAILED_ATTEMPTS:
            new_pin = self._pin_generator.generate()
            reservation.update_pin(new_pin)
            self._repository.save(reservation)
            self._sms_notifier.send(
                reservation.customer.cell_phone,
                f"Your new Wunda Wash PIN is {new_pin:05d}.",
            )
            self._machine_api.lock(
                reservation.id,
                machine_number,
                reservation.slot,
                new_pin,
            )
            self._failure_counts[machine_number] = 0

        return False

File: laundry-reservation/typescript/src/ReservationService.ts
import { randomUUID } from 'node:crypto';
import type { Customer } from './Customer.js';
import type { EmailNotifier } from './EmailNotifier.js';
import { DuplicateReservationError } from './Exceptions.js';
import type { MachineApi } from './MachineApi.js';
import type { MachineSelector } from './MachineSelector.js';
import type { PinGenerator } from './PinGenerator.js';
import { Reservation } from './Reservation.js';
import type { ReservationRepository } from './ReservationRepository.js';
import type { SmsNotifier } from './SmsNotifier.js';

const MAX_FAILED_ATTEMPTS = 5;

function formatPin(pin: number): string {
  return String(pin).padStart(5, '0');
}

export class ReservationService {
  private readonly failureCounts = new Map<number, number>();

  constructor(
    private readonly repository: ReservationRepository,
    private readonly emailNotifier: EmailNotifier,
    private readonly smsNotifier: SmsNotifier,
    private readonly machineApi: MachineApi,
    private readonly pinGenerator: PinGenerator,
    private readonly machineSelector: MachineSelector,
  ) {}

  createReservation(slot: Date, customer: Customer): Reservation {
    const existing = this.repository.findActiveByCustomerEmail(customer.email);
    if (existing) {
      throw new DuplicateReservationError(
        `Customer '${customer.email}' already has an active reservation.`,
      );
    }

    const id = randomUUID();
    const machineNumber = this.machineSelector.selectAvailable();
    const pin = this.pinGenerator.generate();
    const reservation = new Reservation(id, slot, machineNumber, customer, pin);

    this.repository.save(reservation);

    this.emailNotifier.send(
      customer.email,
      'Wunda Wash Reservation Confirmation',
      `Reservation ${id}: Machine ${machineNumber}, PIN ${formatPin(pin)}`,
    );

    this.machineApi.lock(id, machineNumber, slot, pin);

    return reservation;
  }

  claimReservation(machineNumber: number, pin: number): boolean {
    const reservation = this.repository.findActiveByMachineNumber(machineNumber);
    if (!reservation) return false;

    if (reservation.pin === pin) {
      reservation.markUsed();
      this.repository.save(reservation);
      this.machineApi.unlock(machineNumber, reservation.id);
      this.failureCounts.delete(machineNumber);
      return true;
    }

    let failures = (this.failureCounts.get(machineNumber) ?? 0) + 1;
    this.failureCounts.set(machineNumber, failures);

    if (failures >= MAX_FAILED_ATTEMPTS) {
      const newPin = this.pinGenerator.generate();
      reservation.updatePin(newPin);
      this.repository.save(reservation);
      this.smsNotifier.send(
        reservation.customer.cellPhone,
        `Your new Wunda Wash PIN is ${formatPin(newPin)}.`,
      );
      this.machineApi.lock(
        reservation.id,
        machineNumber,
        reservation.slot,
        newPin,
      );
      this.failureCounts.set(machineNumber, 0);
    }

    return false;
  }
}
